neuraxle/pipeline.py

```diff
@@ -92,80 +114,139 @@ def fit(self, data_inputs, expected_outputs=None) -> 'Pipeline':
+            data_container = step.handle_transform(data_container)
 
-        return data_inputs
+        return data_container
 
-    def inverse_transform_processed_outputs(self, processed_outputs) -> Any:
+    def _load_pipeline_checkpoint(self, data_container: DataContainer) -> Tuple[NamedTupleList, DataContainer]:
         """
-        After transforming all data inputs, and obtaining a prediction, we can inverse transform the processed outputs
-        :param processed_outputs: the forward transformed data input
-        :return: backward transformed processed outputs
+        Find the steps left to do, and load the latest checkpoint step data container
+        :param data_container: the data container to resume
+        :return: tuple(steps left to do, last checkpoint data container)
         """
-        for step_name, step in list(reversed(self.steps_as_tuple)):
-            processed_outputs = step.transform(processed_outputs)
-        return processed_outputs
-
-    def read_checkpoint(self, data_inputs) -> Tuple[list, Any]:
-        new_data_inputs = data_inputs
-        new_starting_step_index = self.find_starting_step_index(data_inputs)
+        new_starting_step_index = self._find_starting_step_index(data_container)
 
         step = self.steps_as_tuple[new_starting_step_index]
         if isinstance(step, BaseCheckpointStep):
-            checkpoint = step.read_checkpoint(data_inputs)
-            new_data_inputs = checkpoint
+            data_container = step.read_checkpoint(data_container)
 
-        return self.steps_as_tuple[new_starting_step_index:], new_data_inputs
+        return self.steps_as_tuple[new_starting_step_index:], data_container
 
-    def find_starting_step_index(self, data_inputs) -> int:
+    def _find_starting_step_index(self, data_container: DataContainer) -> int:
         """
-        Find the starting step index that has the latest checkpoint (or 0 if there is no checkpoint)
-        :param data_inputs: the data input to fit on
-        :return: index
+        Find the index of the latest step that can be resumed
+        :param data_container: the data container to resume
+        :return: int index latest resumable step
         """
-        for index, (step_name, step) in enumerate(reversed(self.steps_as_tuple)):
-            if isinstance(step, ResumableStepMixin) and step.should_resume(data_inputs):
+        initial_current_ids = data_container.current_ids
+
+        for index, (step_name, step) in enumerate(self.items()):
+            current_ids = step.hasher.hash(
+                current_ids=data_container.current_ids,
+                hyperparameters=step.hyperparams,
+                data_inputs=data_container.data_inputs
+            )
+            data_container.set_current_ids(current_ids)
+
+        for index, (step_name, step) in enumerate(reversed(self.items())):
+            if isinstance(step, ResumableStepMixin) and step.should_resume(data_container):
```

Comment:
There is a bug here. The last hash is taken only, but each hash of the first for loop must be compared against its respective step. Recursion is needed as follow to keep track of the `new_data_container`: 

```
        for index, (step_name, step) in enumerate(self.items()):
            current_ids = step.hasher.hash(
                current_ids=data_container.current_ids,
                hyperparameters=step.hyperparams,
                data_inputs=data_container.data_inputs
            )
            new_data_container = copy(data_container).set_current_ids(current_ids)
            ret = _find_starting_step_index(new_data_container)
            if ret[0] != -1: 
                return ret
            else:  # Note: I think an `elif` or something is missing here. Coded fast. 
                return -1, ...
```

Reply:
Is this tested? This could be tested with a kind of ReadCheckpointTape and a WriteCheckpointTape class / or some kind of callback classes, like the other tape classes.